File: api/src/main/java/com/example/api/facebook/graphapi/GraphApiBuilder.java
package com.example.api.facebook.graphapi;

import java.io.File;

import okhttp3.Cache;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

public class GraphApiBuilder {
	private static GraphApi graphApi = null;

	public static GraphApi get(Config config, OkHttpClient.Builder okBuilder) {
		if (graphApi == null) {
			graphApi = build(config, okBuilder);
		}

		return graphApi;
	}

	private static GraphApi build(Config config, OkHttpClient.Builder okBuilder) {
		File httpCacheDirectory = new File(config.getCacheDir());
		Cache cache = new Cache(httpCacheDirectory, config.getCacheSize());

		okBuilder.cache(cache);

		return new Retrofit.Builder()
			.baseUrl(config.getUri())
			.addConverterFactory(GsonConverterFactory.create())
			.addCallAdapterFactory(RxJava2CallAdapterFactory.create())
			.client(okBuilder.build())
			.build()
			.create(GraphApi.class);
	}
}

File: libs/api/src/main/kotlin/br/com/vitorsalgado/example/api/graphapi/GraphApiBuilder.kt
package br.com.vitorsalgado.example.api.graphapi

import okhttp3.Cache
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import java.io.File

object GraphApiBuilder {
  private var graphApi: GraphApi? = null

  operator fun get(config: Config, okBuilder: OkHttpClient.Builder): GraphApi {
    if (graphApi == null) {
      graphApi = build(config, okBuilder)
    }

    return graphApi as GraphApi
  }

  private fun build(config: Config, okBuilder: OkHttpClient.Builder): GraphApi {
    val httpCacheDirectory = File(config.cacheDir)
    val cache = Cache(httpCacheDirectory, config.cacheSize.toLong())

    okBuilder.cache(cache)

    return Retrofit.Builder()
      .baseUrl(config.uri)
      .addConverterFactory(GsonConverterFactory.create())
      .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
      .client(okBuilder.build())
      .build()
      .create(GraphApi::class.java)
  }
}
